server/routerlicious/packages/services/src/mongodb.ts

```diff
@@ -414,6 +440,37 @@ export class MongoCollection<T> implements core.ICollection<T>, core.IRetryable
 			}
 		}
 	}
+
+	private terminateBasedOnCounterThreshold(counters: Record<string, number>): void {
+		if (this.apiFailureRateTerminationThreshold > 1) {
+			return; // If threshold set more than 1, meaning we should ever terminate and skip followings.
+		}
+		let totalCount = 0;
+		let totalFailedCount = 0;
+		for (const [apiName, apiCounter] of Object.entries(counters)) {
+			totalCount += apiCounter;
+			if (apiName.endsWith(this.failedApiCounterSuffix)) {
+				totalFailedCount += apiCounter;
+			}
+		}
+
+		if (totalCount < 100) {
+			// too little requests to consider
+			return;
+		}
+		const failureRate = totalFailedCount / totalCount;
+		if (failureRate <= this.apiFailureRateTerminationThreshold) {
+			// OK, no need to terminate
+			return;
+		}
+
+		// This logic is to automate the process of terminates application if db become unfunctional, so
+		// kubernetes would automatically handle the restart process.
+		Lumberjack.warning(
+			`Failure rate ${failureRate} more than threshold ${this.apiFailureRateTerminationThreshold}`,
+		);
+		process.kill(process.pid, "SIGTERM");
```

Comment:
We should probably try to avoid calling process.kill() here. We should try to bubble this up to the RunService infrastructure in here:

https://github.com/microsoft/FluidFramework/blob/077e696729a3bc8abf9b5c69ed3e5fc95f52c3f0/server/routerlicious/packages/services-shared/src/runner.ts#L87

Reasons:
1. I think we should centralize the places in our code that can terminate it. If we start having process.kill() spread throughout the codebase, it will become hard to manage.
2. I suspect that calling process.kill() immediately after the Lumberjack statement might occasionally cause missing logs.
3. The above would cause a service restart but we likely would not have the metrics available for detecting that, since we killed the process would not emit the RunService metric (which happens at the same in the link I shared above)

Reply:
Agree with Henrique about bubbling it up to RunService such that it is caught as an exception [here](https://github.com/microsoft/FluidFramework/blob/077e696729a3bc8abf9b5c69ed3e5fc95f52c3f0/server/routerlicious/packages/services-shared/src/runner.ts#L37). Because [SIGTERM event handler](https://github.com/microsoft/FluidFramework/blob/077e696729a3bc8abf9b5c69ed3e5fc95f52c3f0/server/routerlicious/packages/services-shared/src/runner.ts#L50) calls runner.stop() which doesnt reject the runner promise in case of SIGTERM, and even if the RunService metric is available, it will be hard to identify that it was due to the mongo failure rate.

Reply:
Make sense, however multiple things make me to decide to use `process.kill`:
1. `process.kill` does **NOT** trigger immediately exit (`process.exit` does on the other hand). It will handled by the [SIGTERM event handler](https://github.com/microsoft/FluidFramework/blob/077e696729a3bc8abf9b5c69ed3e5fc95f52c3f0/server/routerlicious/packages/services-shared/src/runner.ts#L50) calls runner.stop() as specified by @shubhi1092 .
2. This layer is actually the deepest layer in our managed code. I can't figure out a decent / (or even workable) way to bubble up an exception from here all the way to reach to the top level layer of `RunService`, not to mention not a single path using this layer, almost all paths are depending on this layer. So I don't think it is possible there is a way for me to be able to bubble up all the way properly for all request paths. 

If you have better ideas how to proper bubble up this, let me know.

Reply:
What happens if we throw an error here? Can we create the error object in a way that is not retried by `requestWithRetry()` in line 374? And if we can do that, does the error bubble up to RunService, or does it get intercepted somewhere else?

Reply:
It will get intercepted. Think about the exception throw here the `MongoServerSelectionTimeout` exception, it if don't get intercepted, it will ends terminates the pods already. However it is not the case for production, so the exception does intercepted by someone somewhere. For example in the meeting you found we reject after 3 retries:
<img width="1721" alt="image" src="https://github.com/microsoft/FluidFramework/assets/3719246/e8ffc959-01e8-4861-a798-42bf0e29c713">

If exception bubbles up properly, the pods would already crashed. Even if we spend efforts on bubble up, it is hard to cover all and future paths given db is the deepest layer. So using signal is actually better sollution.

Reply:
Thanks Xin. I would prefer to not kill the process with process.kill() here. I don't think that is the best approach, and even if we need to do more work to make sure that error bubbles up to RunService, where it would be handled properly, I think that investment is worth it for the reasons I already described above. In fact, I think the purpose of this work item was to do exactly that: find out why the exception is currently being intercepted, and clear the way so that it can finally bubble up to RunService.

If you think that will significantly delay the implementation and cause further issues to the service, then I don't want to slow you down. I will defer that decision to @manishgargd, who might have more insights than me, @shubhi1092 , who recently implemented important improvements in RunService, and to you.

Reply:
@hedasilv @shubhi1092 , @manishgargd  Just rephrase the nodeJS api:
> Even though the name of this function is process.kill(), it is really just a signal sender, like the kill system call. The signal sent may do something other than kill the target process.

It is simple emit an event, to have the listener `RunService` to the termination for us, which is across many many layers and almost ALL code path, it is impractical to bubble up

Reply:
Hey @zhangxin511 , my concern with raising SIGTERM here is - the RunService metric will be logged with successful = true and no information about why RunService was emitted. In other words, you wont be able to identify that the service restarted because of high db failure rate by looking at the RunService metrics. This is because in case of sigterm, [runner.stop() will resolve the runner promise](https://github.com/microsoft/FluidFramework/blob/077e696729a3bc8abf9b5c69ed3e5fc95f52c3f0/server/routerlicious/packages/lambdas-driver/src/kafka-service/runner.ts#L132) and [RunService will be completed as successful](https://github.com/microsoft/FluidFramework/blob/077e696729a3bc8abf9b5c69ed3e5fc95f52c3f0/server/routerlicious/packages/services-shared/src/runner.ts#L112). Because SIGTERM doesnt always mean an exception - eg: manually deleting a pod also triggers a sigterm.

So if your goal is to be able to identify restarts due to high db failure rate, then this will not work - you will need to throw an exception for that. But if you just want to make sure that runner stops gracefully before restarting, then this would work.

Reply:
@shubhi1092 I understand what you are talking about. My point here is an exception throw from db layer, will not reach to the entry point of `RunService`, it will definitely intercepted somewhere, either by code controlled by us, or code controlled by express, that server automatically return 500 to instead of stopping a running application.

Reply:
In that case, we cant use RunService metric to detect this situation, and would need to rely on the kusto logs / some other mechanism for that.

Reply:
This is a good discussion. @zhangxin511 - You have a good point that the exception may be intercepted before it reaches RunService. However, this can be validated in a dev environment, and based on whether it is intercepted or not, we can take a decision.

Reply:
@manishgargd No need to deploy on a dev cluster, I can replicate easily on local env:
I places an exception throw in get delta flow, which is one layer above db:
 
![image](https://github.com/microsoft/FluidFramework/assets/3719246/ed97a774-e613-45e6-82f0-b2821dcfb6c2)

Then request to hit the flow, all get 500:
<img width="470" alt="image" src="https://github.com/microsoft/FluidFramework/assets/3719246/209f2aee-c10a-419b-a669-76e18f348a39">

The alfred didn't stop
![image](https://github.com/microsoft/FluidFramework/assets/3719246/33a57540-2df9-41a1-bf10-46a2636ffe28)


In fact, when rejecting the retry, it is actually throw an exception already, but not necessary to lead to pods stop.

NOTE, this is just one flow of get delta, consider DB layer is used hundreds of flows, how would it be possible to successfully bubble up an exception all the way up to the RunService?